vmm/src/lib.rs

```diff
@@ -1112,6 +1116,26 @@ impl Vmm {
         // Now pause VM
         vm.pause()?;
 
+        #[cfg(target_arch = "aarch64")]
+        {
+            let gic_device = vm.get_gic_device();
+            // Flush redistributors pending tables to guest RAM.
+            save_pending_tables(gic_device.lock().unwrap().device()).map_err(|e| {
+                MigratableError::MigrateSend(anyhow!(
+                    "Error flushing redistributors pending tables: {:#?}",
+                    e
+                ))
+            })?;
+            // Flush ITS tables to guest RAM.
+            gicv3_its_tables_access(gic_device.lock().unwrap().its_device().unwrap(), true)
+                .map_err(|e| {
+                    MigratableError::MigrateSend(anyhow!(
+                        "Error flushing GICv3ITS tables: {:#?}",
+                        e
+                    ))
+                })?;
+        }
```

Comment:
Just a thought but can't we make the GIC device `pause()` implementation taking care of flushing the tables?

Reply:
Yes, I do think this is a good idea. Let me check then.

Reply:
Hi @sboeuf, after checking the current code, I am wondering which method below is better (Could you please kindly share your opinion? Thanks!)

- Flushing the tables in the `pause()` implementation of VM instead of GIC device. This is because currently most of the GIC snapshot/restore/live-migration steps are handled in the VM snapshot/restore/live-migration. The `Pausable` trait of GIC (https://github.com/cloud-hypervisor/cloud-hypervisor/blob/main/arch/src/aarch64/gic/gicv3_its.rs#L504) is left unimplemented.
- Re-implement the `Pausable` trait of GIC and this might need some refactoring.

Reply:
If not too complex, I'd rather see this implemented in the GIC device. But if that's a general trend that everything related to the GIC device is implemented in Vm, then maybe you could do this in two steps. First you implement what you need as part of Vm, and then you create a refactoring PR/patch to move all the migration bits specific to the GIC in the GIC device itself.

Reply:
Ok I will try :)) For me moving it to GIC specific code also makes more sense, but I still need to check how difficult of this refactoring will be. Let's see.